packages/sanity/src/core/releases/tool/detail/useBundleDocuments.ts

```diff
@@ -120,81 +124,165 @@ export function useBundleDocuments(releaseId: string): {
-            }
+          return documentPreviewStore.observeForPreview(document, schemaType).pipe(
+            take(1),
+            map((version) => ({
+              isLoading: false,
+              values: prepareForPreview(
+                getPreviewValueWithFallback({
+                  value: document,
+                  version: version.snapshot || document,
+                  perspective: getReleaseIdFromReleaseDocumentId(release._id),
+                }),
+                schemaType,
+              ),
+            })),
+            startWith({isLoading: true, values: {}}),
+          )
+        }),
+        filter(({isLoading}) => !isLoading),
+      )
 
-            return documentPreviewStore.observeForPreview(document, schemaType).pipe(
-              take(1),
-              // eslint-disable-next-line max-nested-callbacks
-              map((version) => ({
-                isLoading: false,
-                values: prepareForPreview(
-                  getPreviewValueWithFallback({
-                    value: document,
-                    version: version.snapshot || document,
-                    perspective: releaseId,
-                  }),
-                  schemaType,
-                ),
-              })),
-              startWith({isLoading: true, values: {}}),
-            )
-          }),
-          filter(({isLoading}) => !isLoading),
-        )
+      return combineLatest([document$, previewValues$]).pipe(
+        map(([document, previewValues]) => ({
+          document,
+          previewValues,
+          memoKey: uuid(),
+          validation: {validation: [], hasError: false, isValidating: false},
+        })),
+      )
+    }),
+    toArray(),
+    map((results) => ({
+      loading: false,
+      results,
+    })),
+  )
+}
 
-        return combineLatest([document$, previewValues$]).pipe(
-          map(([document, previewValues]) => ({
-            document,
-            previewValues,
-            memoKey: uuid(),
-            validation: {validation: [], hasError: false, isValidating: false},
-          })),
+const getReleaseDocumentsObservable: (
+  observableClient: ObservableSanityClient,
+  schema: Schema,
+  documentPreviewStore: DocumentPreviewStore,
+  dataset: string | undefined,
+  getClient: ReturnType<typeof useSource>['getClient'],
+  releaseId: string,
+  groqFilter: string,
+  i18n: LocaleSource,
+  releasesState$: ReturnType<typeof useReleasesStore>['state$'],
+) => ReleaseDocumentsObservableResult = (
+  observableClient,
+  schema,
+  documentPreviewStore,
+  dataset,
+  getClient,
+  releaseId,
+  groqFilter,
+  i18n,
+  releasesState$,
+  // eslint-disable-next-line max-params
+) =>
+  releasesState$.pipe(
+    map((releasesState) =>
+      releasesState.releases.get(getReleaseDocumentIdFromReleaseId(releaseId)),
+    ),
+    switchMap((release) => {
+      if (!release) return of({loading: true, results: []})
+
```

Comment:
Small suggestion here to update how the missing release is handled and adding a safeguard if something else changes in the releasState$ the operator doesn't start again.
```suggestion
  releasesState$.pipe(
    map((releasesState) =>
      releasesState.releases.get(getReleaseDocumentIdFromReleaseId(releaseId)),
    ),
     filter(Boolean), // Validates release exists
    distinctUntilChanged((prev, next) => prev._rev === next._rev), // Validates the `_rev` has changed, if not, it won't restart the operators.
    switchMap((release) => {
    }),
    startWith({loading: true, results: []}), // Assigns an initial value in case release doesn't exist 


```

Reply:
Thank you - let me have a look through this, I'd forgotten about `distinctUntilChanged`